import gzip
from pysam import VariantFile


def find_snv(file, out_file):
    """
    search for pathogen variant c.607T>C p.(Cys203Arg)

    :param file: vcf.gz file (generated by Clair3)
    :param out_file: txt file for analysis summary
    :return: True if the pathogen variant was found, False otherwise
    """

    # file = "/mnt/storage3b/projects/research/23014I_1145_ONTpilot_TNAMSE/Sample_23014LRa023L2_01/clair_temp
    # /full_alignment.vcf.gz"

    # file = "test_files/sample.compressed.vcf.gz"

    chromosome = "X"
    position = 154154602
    ref_base = "T"
    alt_base = "C"

    vcf = VariantFile(file)

    for record in vcf.fetch('X', position - 1, position):
        if record.pos == position and record.ref == ref_base and alt_base in record.alts:
            print("Pathogen variant c.607T>C p.(Cys203Arg) was found.")

