from pysam import VariantFile
from .variants import SNVs
from .table import write_table


def find_snv(file, table):
    """
    search for SNV's:
    opn1lw: c.607T>C p.(Cys203Arg)
    opn1lw: c.347C>G p.(Ser116Cys)
    opn1lw: c.1013G>T p.(Gly338Val)
    opn1mw: c.607T>C p.(Cys203Arg)
    opn1mw: c.38G>A p.(Arg13His)
    opn1mw: c.659T>C p.(Met220Thr)

    :param file: vcf.gz file (generated by Clair3)
    :param table: table to add found variants
    :return: table with (new) variants
    """

    vcf = VariantFile(file)

    for variant in SNVs:
        chromosome = variant['chromosome']
        position = variant['position']
        ref_base = variant['ref_base']
        alt_base = variant['alt_base']

        for record in vcf.fetch(chromosome, position - 1, position):
            if record.pos == position and record.ref == ref_base and alt_base in record.alts:
                print(f"Variant {variant['variant']} was found.")

                table.append([variant['variant'], variant['Haer-Wigman_classification'], variant['Haer-Wigman_disease'],
                              variant['Neitz_classification'], variant['Neitz_disease']])

    return table